.github/pull_request_template.md

```diff
@@ -0,0 +1,53 @@
+Fixes #XXXX.
+
+(Replace XXXX with the id of the issue fixed in this PR. Remove this line if there is no corresponding
+issue. Don't reference the issue in the title of this pull-request.)
+
+Add tags to your PR if you are a committer (only committers have the right to add tags). Add [Design Review] tag
+if this PR should better be reviewed by at least two people.
+Don't forget to add the following tags (if applicable): [Incompatible], [Release Notes], [Compatibility], [Security],
+[Development Blocker]. Add at least one [Area - ] tag, consider creating a new one if none of the existing [Area - ]
+tags is applicable.
+
+### Description
+
+Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's
+not necessary to repeat the description here, however, you may choose to keep one summary sentence.
+
+Describe your patch: what did you change in code? How did you fix the problem?
+
+If there are several relatively logically separate changes in this PR, list them. For example:
+ - Fixed the bug ...
+ - Renamed the class ...
+ - Added a forbidden-apis entry ...
+
+Some of the aspects mentioned above may be omitted for simple and small PRs.
+
+### Design
+
+Please describe any design decisions made, including:
+ - Choice of algorithms
+ - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such
+   as when insufficient resources are available?
+ - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
+ - Method organization and design (how the logic is split between methods, parameters and return types)
+ - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
+
+In addition, describe _at least one_ alternative design (or mention alternative name) for every design (or naming)
+decision point and compare the alternatives with the designs that you've implemented (or the names you've chosen).
+
+If you already did this in the associated issue (e. g. a "Proposal" issue), leave the following sentence:
+
+Design of this change is discussed [here](<link to Github issue or comment where you discuss the design>).
+
+This section may be omitted for really simple and small patches. However, any patch that adds a new production class
+almost certainly shouldn't omit this section.
+
+<hr>
+
+I've self-reviewed this PR (including using the [concurrency checklist](
+https://github.com/leventov/java-concurrency-checklist)).
```

Comment:
If you think this is really important, I'm fine with adding it. But, this looks biased only toward concurrency issues which is a small part of general programming practices. I feel that it can mislead people to think that we care about only concurrency issues. Why don't we provide other documents as well for more general programming practices?

Reply:
I'm currently relatively actively working on a more general (and more important) checklist about design and documentation, but it may still take a long time until I finish this work. Linking from the PR template to general programming books (even very good ones, like "Effective Java") is not actionable and practical, IMO.

I think it's better to include a link to the concurrency checklist for a while than not include any link at all.

I'll move a link to the concurrency checklist in a separate sentence.

Reply:
I think it'd be better to skip the link completely. The checklist you have posted is long and complex and I am not sure how many Druid committers have read it fully and agree with it. It deserves its own separate discussion.

Reply:
This checklist is ["discussed"](https://docs.google.com/document/d/17EEKT6fih9Dd5NfXjBoECcKbVp1eOB2vb3jKqTF9pPc/edit) [since November](https://lists.apache.org/thread.html/74ec4a52b8cef87129a12d7ee834deef39d8bdef7f049473d4c56058@%3Cdev.druid.apache.org%3E), but nobody participated the discussion except me and @egor-ryashin.

Reply:
I'm suggesting raising a PR containing your checklist to the Druid repo as a way to encourage more discussion.

Reply:
Gian, this is not a good basis for blocking this change because "was not discussed enough", although people had (and still have) all the opportunity to discuss it for 4+ months. I explained why I don't want to copy the checklist into the repo in this comment: https://github.com/apache/incubator-druid/pull/7206#discussion_r264856399

For example, in Apache, when discussing new committers, a discussion where nobody said anything is considered OK and then a vote thread is started anyway.

If you disagree with something in particular in this checklist, you can block this PR template PR for this reason. Then we can discuss the disagreements that should lead to some (or none) overriddings or amendments in the smaller document that I [can create](https://github.com/apache/incubator-druid/pull/7206#discussion_r264856399).

Reply:
To my eyes, the language in the PR right now makes this checklist look like it's being represented as an official Druid style guide. If this is your intent with this PR, I don't think "external document with Druid-specific overrides" is a good approach, for a few reasons:

- It's tough for readers to manually reconcile. We have to ask people to read an external document and mentally apply diffs.
- It's tough for the Druid community to keep updated. We have to maintain a list of diffs rather than an actual document. If the upstream document changes, we have to update our diffs or "pin" to a specific version.
- No other projects seem to use the proposed checklist as part of their official style guide, so there isn't anything to gain from keeping it separate from Druid. Why not make an official Druid checklist and then, if other projects like it and want to use it, they can reference ours?

If your intent was not to represent this as an official Druid checklist, then I don't have as much issue with it, but in that case I think the language should be changed so it doesn't look so official. Maybe refer to this list as one of many possible resources on Java code style, and point out that Druid doesn't have an official coding guide, so contributors are encouraged to try to match the style of other code in the repo. (Basically how it works today.)

Reply:
The idea behind your comment there will be some diffs. Actually, I don't think there will be anything in the Druid's document that is *opposite* to what the general checklist says. Because if the necessity for such points araises, then the general checklist is not general enough and the controversial points need to be softened or removed.

So, in fact, the Druid's document will contain only amendments (such as "Is it possible to use `LifecycleLock` instead of intrinsic lock for a lifecycled object?") or refinements, but not overridings. The users won't need to mentally apply diffs.

Reply:
I assume there will be diffs at some point, since style guides aren't set in stone.

Reply:
This is not a style (like, opinionated) checklist. Generality is really its goal. For example, pretty much all concurrency experts that I consulted with said that they have rules in their projects to *prohibit* daemon threads, but because Druid has the opposite policy, I've worded the [corresponding item](https://github.com/code-review-checklists/java-concurrency#name-threads) to not give preference to any approach.